### PR TITLE
many: update apparmor to 4.0.1

### PR DESCRIPTION
**Pull request by zyga**
This is a rebase of https://github.com/snapcore/snapd/pull/13354

This is replacement for older attempt by Alex M: https://github.com/snapcore/snapd/pull/13354

Bug reports and fixes to consider: https://ubuntu-archive-team.ubuntu.com/pending-sru.html
- [2046844](https://launchpad.net/bugs/2046844) 
- [2056297](https://launchpad.net/bugs/2056297) 
- [2060100](https://launchpad.net/bugs/2060100) 
- [2064672](https://launchpad.net/bugs/2064672) 
- [2065708](https://launchpad.net/bugs/2065708)

### CODE REVIEW

**Comment by alexmurray** [comment]
Thanks @zyga - so like the original PR, we still see failures in `tests/main/snapd-homedirs-vendored` which I also explained back in the original PR - https://github.com/snapcore/snapd/pull/13354#issuecomment-2011246767 - I am not sure there is a good fix for this - every time we update to a newer vendored AppArmor which has a new default ABI, this test will likely fail in the same way due to the mismatch between the old and new ABIs. But once this is merged to master then the test will pass 🙃

File: build-aux/snap/local/apparmor/af_names.h
Comment (valentindavid) [inline build-aux/snap/local/apparmor/af_names.h:2]: We should probably remove this file once we have moved to using core24 as build base. Right?

**Review by valentindavid — APPROVED** [review]
LGTM

File: build-aux/snap/local/apparmor/af_names.h
Comment (zyga) [reply build-aux/snap/local/apparmor/af_names.h:2]: This is a bit of an annoying part. The parser looks at the host's socket.h file to determine what kind of sockets one can create and to know their names.

If the snap is built on an old system, where the definition of old is always inaccurate. In theory we could drop it but we'd alwayd be at the risk of compiling but silently failing to understand specific network types.

**Review by jslarraz — APPROVED** [review]
LGTM

**Review by ernestl — COMMENTED** [review]
Some comments/questions for consideration

File: sandbox/apparmor/apparmor.go
Comment (zyga) [reply sandbox/apparmor/apparmor.go:817]: This comes straight from Alex. As for stability, it's as stable as those things get (no promise, usually okay).

File: sandbox/apparmor/apparmor.go
Comment (zyga) [reply sandbox/apparmor/apparmor.go:1]: Can you clarify the 2nd link to a diff that's about not covered by test? Either the link is wonky or I don't understand what the question is.

I'll look into 4.0 testing in snapd-apparmor, thanks!

File: sandbox/apparmor/apparmor.go
Comment (ernestl) [reply sandbox/apparmor/apparmor.go:817]: ok

File: sandbox/apparmor/apparmor.go
Comment (zyga) [reply sandbox/apparmor/apparmor.go:1]: I've made a separate patch to enable apparmor 4.0 ABI when using the host compiler. I do not want to introduce it here as it's something we cannot easily synchronize with. It should be considered and reviewed separately. I also write some tests in the same pass.

The follow-up is: https://github.com/snapcore/snapd/pull/14167

**Review by pedronis — APPROVED** [review]
+1

**Review by ernestl — APPROVED** [review]
Thanks

**Comment by zyga** [comment]
There's a number of failures that need investigation. Please don't merge this yet @ernestl

**Comment by zyga** [comment]
This is hitting the /dev/mqueue issue with apparmor on the host.

**Comment by zyga** [comment]
AppArmor 4.0.1 SRU has been released to Noble.

**Comment by zyga** [comment]
Various tests are now failing on:
```
+ snap debug sandbox-features --required apparmor:parser:snapd-internal
error: sandbox feature not available: "apparmor:parser:snapd-internal"
```

I have a hunch that this is because apparmor on the host and in snapd is the same version. Perhaps we need to fake the version to "+1" or fix the tests to cope?

**Comment by zyga** [comment]
The test `google:ubuntu-18.04-64:tests/main/snapd-homedirs-vendored` fails as in the original branch from Alex, due to the fact that when using apparmor re-execution we assume 4.0 is present and in that specific case, it is not as we repackage snapd snap from the store without injecting updated apparmor.

**Comment by zyga** [comment]
I've improved the code a little so that tests might actually pass now.

**Comment by zyga** [comment]
If tests are happy this needs a re-review.

**Comment by zyga** [comment]
Dear reviewers, please see patches:

- https://github.com/snapcore/snapd/pull/14150/commits/700f794b5a97ac377bee3a2abd2749d183588ebf 
- https://github.com/snapcore/snapd/pull/14150/commits/c5df0af3e94018de9747321bddaa0468838422c3

Those are non-test changes that have happened since the branch was mainly approved.

**Review by ernestl — APPROVED** [review]
Thanks, nice upgrade with ability to now manage version specific feature bugs and also the ability to use available internal ABI file!

Some nitpick comments for consideration.

**Pull request merged**
